tests/Aspire.Hosting.Tests/Dapr/DaprTests.cs

```diff
@@ -77,10 +77,96 @@ public async Task WithDaprSideCarAddsAnnotationAndSidecarResource()
-            "localhost"
+            "localhost",
+            "--app-protocol",
+            "http"
         };
 
         Assert.Equal(expectedArgs, sidecarArgs);
         Assert.NotNull(container.Annotations.OfType<DaprSidecarAnnotation>());
     }
+
+    [Theory]
+    [InlineData("https", "https", 555, "https", "localhost", 555)]
+    [InlineData(null, null, null, "http", "localhost", 8000)]
+    [InlineData("https", null, null, "https", "localhost", 8001)]
+    [InlineData(null, "https", null, "https", "localhost", 8001)]
+    [InlineData(null, null, 555, "http", "localhost", 555)]
+    [InlineData("https", "http", null, "https", "localhost", 8000)]
+    public async Task WithDaprSideCarAddsAnnotationBasedOnTheSidecarAppOptions(string? schema, string? endPoint, int? port, string expectedSchema, string expectedChannelAddress, int expectedPort)
+    {
+        using var builder = TestDistributedApplicationBuilder.Create(new DistributedApplicationOptions
+        {
+            DisableDashboard = true
+        });
+
+        builder.AddDapr(o =>
+        {
+            // Fake path to avoid throwing
+            o.DaprPath = "dapr";
+        });
+
+        var containerResource = builder.AddContainer("name", "image")
+            .WithEndpoint("http", e =>
+            {
+                e.Port = 8000;
+                e.UriScheme = "http";
+                e.AllocatedEndpoint = new(e, "localhost", 8000);
+            })
+            .WithEndpoint("https", e =>
+            {
+                e.Port = 8001;
+                e.UriScheme = "https";
+                e.AllocatedEndpoint = new(e, "localhost", 8001);
+            });
+        if (schema is null && endPoint is null && port is null)
+        {
+            containerResource.WithDaprSidecar();
+        }
+        else
+        {
+            containerResource.WithDaprSidecar(new DaprSidecarOptions()
+            {
+                AppProtocol = schema,
+                AppEndpoint = endPoint,
+                AppPort = port
+            });
+        }
+        using var app = builder.Build();
+        await app.ExecuteBeforeStartHooksAsync(default);
+
+        var model = app.Services.GetRequiredService<DistributedApplicationModel>();
+
+        Assert.Equal(3, model.Resources.Count);
+        var container = Assert.Single(model.Resources.OfType<ContainerResource>());
+        var sidecarResource = Assert.Single(model.Resources.OfType<IDaprSidecarResource>());
+        var sideCarCli = Assert.Single(model.Resources.OfType<ExecutableResource>());
+
+        Assert.True(sideCarCli.TryGetEndpoints(out var endpoints));
+
+        var ports = new Dictionary<string, int>
+        {
+            ["http"] = 3500,
+            ["grpc"] = 50001,
+            ["metrics"] = 9090
+        };
+
+        foreach (var e in endpoints)
+        {
+            e.AllocatedEndpoint = new(e, "localhost", ports[e.Name], targetPortExpression: $$$"""{{- portForServing "{{{e.Name}}}" -}}""");
```

Comment:
Is it necessary to set the `targetPortsExpression` here?

Reply:
I'm not sure if the dapr related code would work without it. I will take a look later.

Reply:
@karolz-ms  yes it is important. If you don't do that you will get the following exception inside of `Aspire.Hosting.ApplicationModel.EndpointReference`

![image](https://github.com/dotnet/aspire/assets/11291885/abc23dc5-152e-43af-a392-051c08c6c774)

Reply:
OK thanks for confirmation